pgbouncer/kubedb_client_builder.go

```diff
@@ -0,0 +1,175 @@
+)
+
+type KubeDBClientBuilder struct {
+	kc          client.Client
+	db          *api.PgBouncer
+	url         string
+	podName     string
+	pgBouncerDB string
+}
+
+func NewKubeDBClientBuilder(kc client.Client, db *api.PgBouncer) *KubeDBClientBuilder {
+	return &KubeDBClientBuilder{
+		kc: kc,
+		db: db,
+	}
+}
+
+func (o *KubeDBClientBuilder) WithPod(podName string) *KubeDBClientBuilder {
+	o.podName = podName
+	return o
+}
+
+func (o *KubeDBClientBuilder) WithURL(url string) *KubeDBClientBuilder {
+	o.url = url
+	return o
+}
+
+func (o *KubeDBClientBuilder) WithPgBouncerDB(pgDB string) *KubeDBClientBuilder {
+	o.pgBouncerDB = pgDB
+	return o
+}
+
+func (o *KubeDBClientBuilder) GetPgbouncerXormClient() (*XormClient, error) {
+	connector, err := o.getConnectionString()
+	if err != nil {
+		return nil, err
+	}
+
+	engine, err := xorm.NewEngine("postgres", connector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate pgbouncer client using connection string: %v", err)
+	}
+	_, err = engine.Query("SHOW HELP")
+	if err != nil {
+		return nil, fmt.Errorf("failed to run query: %v", err)
+	}
+	return &XormClient{engine}, nil
+}
+
+func (o *KubeDBClientBuilder) getURL() string {
+	return fmt.Sprintf("%s.%s.%s.svc", o.podName, o.db.GoverningServiceName(), o.db.Namespace)
+}
+
+func (o *KubeDBClientBuilder) getPgBouncerAuthCredentials() (string, string, error) {
+	if o.db.Spec.AuthSecret == nil {
+		return "", "", errors.New("no database secret")
+	}
+	var secret core.Secret
+	err := o.kc.Get(context.TODO(), client.ObjectKey{Namespace: o.db.Namespace, Name: o.db.Spec.AuthSecret.Name}, &secret)
+	if err != nil {
+		return "", "", err
+	}
+	return string(secret.Data[core.BasicAuthUsernameKey]), string(secret.Data[core.BasicAuthPasswordKey]), nil
+}
+
+func (o *KubeDBClientBuilder) GetPgBouncerClient() (*Client, error) {
+	connector, err := o.getConnectionString()
+	if err != nil {
+		return nil, err
+	}
+	// connect to database
+	db, err := sql.Open("postgres", connector)
+	if err != nil {
+		return nil, err
+	}
+
+	// ping to database to check the connection
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	if _, err := db.QueryContext(ctx, "SHOW HELP;"); err != nil {
+		klog.Infoln(db)
+		klog.Infoln("here is error", connector)
+		log.Fatal(err)
+	}
+
+	return &Client{db}, nil
+}
+
+func (o *KubeDBClientBuilder) getConnectionString() (string, error) {
+	if o.podName != "" {
+		o.url = o.getURL()
+	}
+	dnsName := o.url
+	port := 5432
+
+	if o.pgBouncerDB == "" {
+		o.pgBouncerDB = DefaultPgBouncerDB
+	}
+
+	user, pass, err := o.getPgBouncerAuthCredentials()
+	if err != nil {
+		return "", fmt.Errorf("DB basic auth is not found for PgBouncer %v/%v", o.db.Namespace, o.db.Name)
+	}
+	cnnstr := ""
+	sslMode := o.db.Spec.SSLMode
+	if sslMode == "" {
+		sslMode = api.PgBouncerSSLModeDisable
+	}
+
+	if o.db.Spec.TLS != nil {
+		secretName := o.db.GetCertSecretName(api.PgBouncerClientCert)
```

Comment:
can we refactor this part into a different method ? Like `func (o *KubeDBClientBuilder) getTLSConfig() (tlsConfig, error)`. Then use it in building connection string.